public/app/features/alerting/unified/utils/rulerClient.ts

```diff
@@ -157,65 +141,106 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
-        await setRulerRuleGroup(rulerConfig, freshExisting.namespace, {
-          name: freshExisting.group.name,
-          interval: evaluateEvery,
-          rules: [formRule],
-        });
-        return { uid, ruleSourceName: 'grafana' };
-      }
+    if (!existingRule) {
+      return addRuleToNamespaceAndGroup(namespace, groupSpec, newRule);
     }
 
-    // if creating new rule or folder was changed, create rule in a new group
-    const targetFolderGroups = await fetchRulerRulesNamespace(rulerConfig, folder.title);
+    const sameNamespace = existingRule.namespace === namespace;
+    const sameGroup = existingRule.group.name === values.group;
+    const sameLocation = sameNamespace && sameGroup;
+
+    if (sameLocation) {
+      // we're update a rule in the same namespace and group
+      return updateGrafanaRule(existingRule, newRule, evaluateEvery);
+    } else {
+      // we're moving a rule to either a different group or namespace
+      return moveGrafanaRule(namespace, groupSpec, existingRule, newRule);
+    }
+  };
 
-    // set group name to rule name, but be super paranoid and check that this group does not already exist
-    const groupName = getUniqueGroupName(values.name, targetFolderGroups);
-    formRule.grafana_alert.title = groupName;
+  // add a rule to a namespace and group
+  const addRuleToNamespaceAndGroup = async (
+    namespace: string,
+    group: { name: string; interval: string },
+    newRule: PostableRuleGrafanaRuleDTO
+  ): Promise<RuleIdentifier> => {
+    const existingGroup = await fetchRulerRulesGroup(rulerConfig, namespace, group.name);
+    if (!existingGroup) {
+      throw new Error(`No group found with name "${group.name}"`);
+    }
 
     const payload: PostableRulerRuleGroupDTO = {
-      name: groupName,
-      interval: evaluateEvery,
-      rules: [formRule],
+      name: group.name,
+      interval: group.interval,
+      rules: (existingGroup.rules ?? []).concat(newRule as RulerGrafanaRuleDTO),
     };
-    await setRulerRuleGroup(rulerConfig, folder.title, payload);
 
-    // now refetch this group to get the uid, hah
-    const result = await fetchRulerRulesGroup(rulerConfig, folder.title, groupName);
-    const newUid = (result?.rules[0] as RulerGrafanaRuleDTO)?.grafana_alert?.uid;
-    if (newUid) {
-      // if folder has changed, delete the old one
-      if (existing) {
-        const freshExisting = await findEditableRule(ruleId.fromRuleWithLocation(existing));
-        if (freshExisting && freshExisting.namespace !== folder.title) {
-          await deleteRule(freshExisting);
-        }
-      }
+    await setRulerRuleGroup(rulerConfig, namespace, payload);
 
-      return { uid: newUid, ruleSourceName: 'grafana' };
-    } else {
-      throw new Error('Failed to fetch created rule.');
+    // TODO figure out how to get the newly created UID :(
+    return { uid: '', ruleSourceName: GRAFANA_RULES_SOURCE_NAME };
+  };
+
+  // we can't move the rule atomically so we have to
```

Comment:
nit: spelling

Reply:
I meant so clarify here that there isn't a single atomic operation we can perform (atomically), but I can change the wording :D